internal/services/compute/managed_disk_resource.go

```diff
@@ -298,12 +306,25 @@ func resourceManagedDiskCreate(d *pluginsdk.ResourceData, meta interface{}) erro
 	if createOption == compute.DiskCreateOptionFromImage {
 		imageReferenceId := d.Get("image_reference_id").(string)
 		if imageReferenceId == "" {
-			return fmt.Errorf("`image_reference_id` must be specified when `create_option` is set to `Import`")
+			return fmt.Errorf("`image_reference_id` must be specified when `create_option` is set to `FromImage`")
 		}
 
-		props.CreationData.ImageReference = &compute.ImageDiskReference{
+		imageDiskReference := &compute.ImageDiskReference{
 			ID: utils.String(imageReferenceId),
 		}
+
+		imageReferenceLun := d.Get("image_reference_lun").(int)
+		if imageReferenceLun != -1 {
```

Comment:
is there a reason we are not using `d.GetOk` ?

Reply:
The valid values for `image_reference_lun` are 0, 1, 2, 3, ..., which is the logic unit number of the data disk. In order to make 0 a valid value, I used `-1` as its default value. `d.GetOk` is not able to check the -1 so I used `d.Get` instead.

Reply:
Hi @myc2h6o - `d.GetOk()` will pick up if the value has been set by the user, so you can remove the `Default` schema property, and this should should allow the value `0`.
Which should let you change this to something like:
```suggestion
		if imageReferenceLun, ok := d.GetOk("image_reference_lun").(int); ok {
```

Reply:
I removed the `Default` schema property and used `d.GetOk`, however the `ok` is `false` when I pass `0` in the config. This method seems to [check whether the value is zero](https://github.com/hashicorp/terraform-plugin-sdk/blob/76b3563a7afe40aa29d08d1661c5502b59e3811a/helper/schema/resource_data.go#L88) besides checking if it's set by user.
This property is a little bit tricky. When not present in api, it means os disk, when present, it means index of the data disk and can be set to 0 because the index of data disk starts from 0. In order to include both os disk and data disk, I used -1 to mean os disk, and value >= 0 to mean data disk

Reply:
What's the use-case for adding this field to Terraform? Presumably this field will be set/conflict when a Disk is added from a virtual machine, so whilst I can see the field exists in the API - I have a feeling exposing this may cause larger issues?

Reply:
yes it makes more sense to attach a clean data disk to vm using `azurerm_virtual_machine_data_disk_attachment`. I've removed the disk lun property